[빙봉] API 테스트/문서자동화 미션 제출합니다.

aegis1920 opened the pull request:
또 뵙네요 후니! 많은 피드백 부탁드립니다 :)

aegis1920 committed:
feat : docs를 만드는 기능 추가

jjj0611 committed:
[feat] basic http authentication 학습 테스트 및 구현 완료

aegis1920 committed:
Merge branch 'jjj0611' of https://github.com/jjj0611/atdd-subway-favorite into bb-practice

aegis1920 committed:
[feat] session http authentication 학습 테스트 및 구현 완료

aegis1920 committed:
[feat] bearer http authentication 학습 테스트 및 구현 완료

aegis1920 committed:
docs : README.md 추가 및 기본 요구사항 정리

jjj0611 committed:
feat : join 기능과 페이지 연동

	- Join.js 파일 생성으로 /members server와 front 연결

aegis1920 committed:
feat : Email을 중복 체크하는 기능

- MemberController에서 도메인을 쓰지 않도록 수정
- MemberService에서 도메인을 쓰도록 수정

jjj0611 committed:
feat : 회원가입 기능과 로그인 기능을 front와 연동

	- bearer authentication 방식을 사용하여 로그인 기능 구현
	- authorization code는 local storage에 저장
	- 사용하지 않는 session과 basic interceptor 삭제

aegis1920 committed:
docs : 사용자 CRUD 인수테스트 업데이트

aegis1920 committed:
feat : Member CRUD 인수테스트 및 인증 인터셉터 구현

- MemberAuthInterceptor 생성

jjj0611 committed:
refactor : 로그인된 사용자의 정보에 관한 요청을 LoginMemberController로 변경

jjj0611 committed:
docs : 즐겨찾기 기능 인수 테스트 README.md 추가

jjj0611 committed:
feat : 즐겨 찾기 기능에 대한 인수테스트 추가 및 Mock 서버 구현

	- 즐겨찾기 기능 인수테스트 추가
	- 인수테스트를 통과시키기 위한 Mock 서버 구현

aegis1920 committed:
feat : Favorite CRUD 작성

- FavoriteService 테스트 코드 작성
- FavoriteRepository 작성
- FavoriteController 구현
- FavoriteService 구현

aegis1920 committed:
feat : 내 정보 보기 및 수정 페이지 연동

jjj0611 committed:
feat : 즐겨찾기 api와 페이지 연동

	- 즐겨찾기 추가 기능 연동
	- 즐겨찾기 삭제 기능 연동
	- 즐겨찾기 목록 조회 기능 연동

aegis1920 committed:
docs : Member API Documentation 자동화

jjj0611 committed:
docs : favorite api 문서 생성

	- favorite api 문서 생성을 위한 controller 테스트 추가 및 문서 생성
	- add favorite 및 get favorite에 대한 문서 생성

aegis1920 committed:
feat : 즐겨찾기 목록 불러오기 Controller 테스트 작성 및 API 문서 자동화

jjj0611 committed:
docs : Favorite controller remove api 문서 작성

	- favorite api remove api 문서 작성을 위한 테스트 추가

aegis1920 committed:
docs : favorite controller deleteById 테스트 코드 작성 및 API 문서화

aegis1920 committed:
docs : LoginMember API 문서화

        - LoginMemberController 테스트 작성
        - Etag 테스트 수정
        - LoginMemberController API 문서화

aegis1920 committed:
feat : 에러를 보여주는 기능 추가

    - 사용하지 않는 메서드 삭제
    - requestDTO에 Validate하는 어노테이션 추가
    - Controller에 @Valid 어노테이션 추가
    - favoriteDTO를 dto 패키지로 이동
    - ErrorResponse 및 SameStation에 대한 validate 어노테이션 추가

jeonghoon1107 commented on src/test/java/wooteco/subway/web/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,156 @@
+@ExtendWith(RestDocumentationExtension.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+public class FavoriteControllerTest {
```
문서가 잘 만들어지나요?
깨지는 테스트에 대해서 모두 수정해주세요 :)

aegis1920 replied:
현재 제 컴퓨터에서는 53개의 테스트 모두 통과되고 문서도 잘 만들어집니다.
어떤 부분이 깨지는 지 말씀해주실 수 있을까요?

jeonghoon1107 replied:
제 노트북 상태가 이상했나보네요 ㅠㅠ 잘 됩니다 👍

jeonghoon1107 commented on src/main/java/wooteco/subway/web/member/MemberController.java:
```diff
@@ -1,15 +1,20 @@
-
 @RestController
 public class MemberController {
     private MemberService memberService;
```
final로 받아주세요 :)
자바에서 final이 어떤 의미로 사용되는지 찾아보시면 좋을 것 같아요.

aegis1920 replied:
수정했습니다 :)

final을 사용하면 재할당을 금지해 immutable하게 만들 수 있습니다. immutable하게 만들면 변경이 되지 않아 thread safe하며, 안전한 재사용이 가능합니다.

Controller에 있는 service 필드의 경우, 생성자 주입 방식을 통해 DI를 하도록 했습니다.
이는 bean(싱글 오브젝트)으로 만들어 객체를 재사용할 수 있도록 합니다.

bean은 매번 재사용되는 싱글 오브젝트이기 때문에 변경될 일이 없으므로 final을 사용하는 것이 더욱 좋은 클린 코드라고 생각합니다.

제가 제대로 이해한 게 맞을까요?

jeonghoon1107 replied:
final로 선언하면 반드시 생성자를 통해야하기 때문에 컴파일 시점에 생성자로 주입되지 않은 것들에 대해 알 수 있겠네요!

jeonghoon1107 commented on src/main/java/wooteco/subway/infra/JwtTokenProvider.java:
```diff
@@ -1,18 +1,24 @@
 
-    public JwtTokenProvider(@Value("${security.jwt.token.secret-key}") String secretKey, @Value("${security.jwt.token.expire-length}") long validityInMilliseconds) {
+    public JwtTokenProvider(@Value("${security.jwt.token.secret-key}") String secretKey,
+        @Value("${security.jwt.token.expire-length}") long validityInMilliseconds) {
```
https://www.baeldung.com/spring-enable-config-properties
적용해보면 좋을 것 같아요.
장단점이 무엇인지도 함께 공부해보세요!

aegis1920 replied:
적용할 값이 secret-key와 expire-length로 2개 이기에 @ConfigurationProperties를 적용했습니다 :)

`@ConfigurationProperties`

- 호출해서 사용하려는 Properties 정보가 여러 개라면 Bean으로 등록해 사용
- 객체에 바인딩 해줄 수 있음

`@Value`

- 단일 정보만 사용한다면 @Value를 사용
- 단점으로는 키 값이 완전히 똑같은 순수한 포맷으로 바인딩해야함

jeonghoon1107 commented on src/main/java/wooteco/subway/web/member/LoginMemberController.java:
```diff
@@ -20,26 +25,27 @@ public LoginMemberController(MemberService memberService) {
-    public ResponseEntity<MemberResponse> getMemberOfMineBasic(@LoginMember Member member) {
-        return ResponseEntity.ok().body(MemberResponse.of(member));
+    @DeleteMapping("/me")
+    public ResponseEntity<Void> deleteMe(@LoginMemberId Long id) {
```
멤버가 탈퇴하면 favorite은 어떻게 되나요?

aegis1920 replied:
같이 삭제되도록 수정했습니다! 지적 감사합니다 :)

jeonghoon1107 reviewed (changes requested):
안녕하세요 빙봉!
깔끔하게 구현 잘 하셨네요 👍 
피드백을 남겼으니 확인해주세요 :)

jeonghoon1107 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,76 @@
+            .build();
+    }
+
+    @GetMapping(value = "/favorites/{source}/{target}")
```
https://meetup.toast.com/posts/92
참고해보시면 좋을 것 같아요 :)

aegis1920 replied:
여러 즐겨찾기에서 하나의 출발점과 하나의 도착점이 있다고 생각해 `"/favorites/source/{source}/target/{target}"`으로 작성해보았습니다 :)

aegis1920 committed:
refactor : 필드에 final 추가

aegis1920 committed:
refactor : @Value에서 @ConfigurationProperites로 적용

aegis1920 committed:
refactor : 멤버가 삭제되면 즐겨찾기도 삭제되도록 수정

aegis1920 committed:
refactor : /favorites/source/target URI 변경

jeonghoon1107 commented on src/main/java/wooteco/subway/config/WebMvcConfig.java:
```diff
@@ -1,37 +1,30 @@
+        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/favorites/**");
     }
 
     @Override
```
타입은 명확하게 해주세요 :)

aegis1920 replied:
제네릭 타입을 추가해보았습니다!

jeonghoon1107 commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public void removeFavorite(Long memberId, Long source, Long target) {
+        boolean ok = favoriteRepository.deleteByMemberIdAndSourceAndTarget(memberId, source,
```
변수명을 명확하게 써주세요 :)

aegis1920 replied:
`isDeleted`라는 변수명을 사용해봤습니다!

jeonghoon1107 commented on src/main/java/wooteco/subway/web/member/LoginMemberMethodArgumentResolver.java:
```diff
@@ -1,41 +1,29 @@
-        } catch (Exception e) {
-            throw new InvalidAuthenticationException("비정상적인 로그인");
+        NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
+        String id = (String)webRequest.getAttribute("loginMemberId", SCOPE_REQUEST);
```
Interceptor에서도 사용하는 것 같은데 loginMemberId를 상수로 관리하는 것이 어떨까요?

aegis1920 replied:
LoginMemberMethodArgumentResolver안에서 상수로 바꿔보았습니다!

jeonghoon1107 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,76 @@
+        return ResponseEntity.ok(favoriteResponse);
+    }
+
+    @GetMapping("/favorites")
```
마치 모든 favorites를 가져올 것 같은 path네요!

aegis1920 replied:
자신의 모든 즐겨찾기를 가지고 오는데 어차피 자신만 볼 수 있으니 간단하게 `/favorites`만 적어도 될 것 같다고 생각했습니다. 그러나 자신이 가지고 있는 즐겨찾기이므로 `/me/favorites`라고 적는 게 더 명확할 것 같아 `/me/favorites`로 수정해보았습니다!

jeonghoon1107 commented on src/main/java/wooteco/subway/web/member/LoginMemberMethodArgumentResolver.java:
```diff
@@ -1,41 +1,29 @@
+        NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
+        String id = (String)webRequest.getAttribute("loginMemberId", SCOPE_REQUEST);
+        if (StringUtils.isBlank(id)) {
+            return null;
```
비어있으면 잘못된 잘못된 토큰을 받은 것 아닌가요? 
적절한 status와 예외를 주면 어떨까요?

aegis1920 replied:
InvalidAuthenticationException을 throw 하도록 했습니다!

jeonghoon1107 commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthInterceptor.java:
```diff
@@ -1,48 +1,52 @@
 
-        // TODO: 추출한 토큰값의 유효성 검사 (jwtTokenProvider.validateToken() 메서드 활용)
+        if (!jwtTokenProvider.validateToken(accessToken)) {
+            throw new IllegalArgumentException("유효하지 않은 토큰입니다!");
```
토큰이 유효하지 않으면 500 에러인가요?

aegis1920 replied:
인증에 대한 부분이므로 ControllerAdvice에서 401 http code를 보내도록 했습니다!

jeonghoon1107 reviewed (changes requested):
안녕하세요 빙봉!
간단한 피드백남겼으니 확인부탁드려요 :)

aegis1920 committed:
refactor : 명확한 제네릭 타입으로 수정 및 변수명 수정

aegis1920 committed:
refactor : loginMemberId 상수화

aegis1920 committed:
refactor : favorites URI를 /me/favorites로 수정

aegis1920 committed:
refactor : 인증되지 않은 요청일 때 에러 처리 수정

- 인증되지 않은 요청일 경우, InvalidAuthenticationException을
  throw하도록 수정
- ControllerAdvice에서 InvalidAuthenticationException일 경우, 401을
  보내도록 수정

jeonghoon1107 commented on src/test/java/wooteco/subway/service/favorite/FavoriteServiceTest.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @Test
+    void getFavoritesTest() {
```
mock을 하지 않고 h2 db를 이용해 테스트해보면 어떨까요?

jeonghoon1107 reviewed (changes requested):
안녕하세요 빙봉!
깔끔하네요 👍 
간단한 피드백을 남겼으니 확인해주세요 :)

aegis1920 committed:
refactor : mock을 사용하지 않고 h2 db를 사용하도록 함

- FavoriteServiceTest
- MemberServiceTest

jeonghoon1107 reviewed (approved):
안녕하세요 빙봉!
미션 진행하느라 고생하셨습니다 :)
